File: modules/react-native-mlkit-barcode-scanning/ios/RNMLKitBarcodeScannerResult.swift
//
//  RNMLKitBarcodeScannerResult.swift
//  RNMLKitBarcodeScanning
//

import ExpoModulesCore
import Foundation
import MLKitBarcodeScanning
import MLKitVision

// MARK: - Main Result

struct RNMLKitBarcodeScannerResultRecord: Record {
    @Field
    var barcodes: [RNMLKitBarcode] = []
}

public class RNMLKitBarcodeScannerResult {
    var barcodes: [MLKitBarcodeScanning.Barcode]

    init(barcodes: [MLKitBarcodeScanning.Barcode]) {
        self.barcodes = barcodes
    }

    var record: RNMLKitBarcodeScannerResultRecord {
        let result = RNMLKitBarcodeScannerResultRecord()
        result.barcodes = self.barcodes.map { barcode in
            RNMLKitBarcode.from(barcode: barcode)
        }
        return result
    }
}

// MARK: - Barcode

struct RNMLKitBarcode: Record {
    @Field
    var frame: RNMLKitRect = .init()
    @Field
    var rawValue: String?
    @Field
    var rawData: [Int]?
    @Field
    var displayValue: String?
    @Field
    var format: String = ""
    @Field
    var cornerPoints: [RNMLKitPoint] = []
    @Field
    var valueType: String = ""
    @Field
    var email: RNMLKitBarcodeEmail?
    @Field
    var phone: RNMLKitBarcodePhone?
    @Field
    var sms: RNMLKitBarcodeSMS?
    @Field
    var url: RNMLKitBarcodeURLBookmark?
    @Field
    var wifi: RNMLKitBarcodeWifi?
    @Field
    var geoPoint: RNMLKitBarcodeGeoPoint?
    @Field
    var contactInfo: RNMLKitBarcodeContactInfo?
    @Field
    var calendarEvent: RNMLKitBarcodeCalendarEvent?
    @Field
    var driverLicense: RNMLKitBarcodeDriverLicense?

    static func from(barcode: MLKitBarcodeScanning.Barcode) -> RNMLKitBarcode {
        var result = RNMLKitBarcode()

        result.frame = RNMLKitRect.fromCGRect(rect: barcode.frame)
        result.rawValue = barcode.rawValue
        result.rawData = barcode.rawData.map { [UInt8]($0).map { Int($0) } }
        result.displayValue = barcode.displayValue
        result.format = formatToString(barcode.format)
        result.cornerPoints = barcode.cornerPoints?.map { point in
            RNMLKitPoint.fromNSValue(point: point)
        } ?? []
        result.valueType = valueTypeToString(barcode.valueType)

        // Convert nested objects if present
        if let email = barcode.email {
            result.email = RNMLKitBarcodeEmail.from(email: email)
        }

        if let phone = barcode.phone {
            result.phone = RNMLKitBarcodePhone.from(phone: phone)
        }

        if let sms = barcode.sms {
            result.sms = RNMLKitBarcodeSMS.from(sms: sms)
        }

        if let urlBookmark = barcode.url {
            result.url = RNMLKitBarcodeURLBookmark.from(url: urlBookmark)
        }

        if let wifi = barcode.wifi {
            result.wifi = RNMLKitBarcodeWifi.from(wifi: wifi)
        }

        if let geoPoint = barcode.geoPoint {
            result.geoPoint = RNMLKitBarcodeGeoPoint.from(geoPoint: geoPoint)
        }

        if let contactInfo = barcode.contactInfo {
            result.contactInfo = RNMLKitBarcodeContactInfo.from(contactInfo: contactInfo)
        }

        if let calendarEvent = barcode.calendarEvent {
            result.calendarEvent = RNMLKitBarcodeCalendarEvent.from(calendarEvent: calendarEvent)
        }

        if let driverLicense = barcode.driverLicense {
            result.driverLicense = RNMLKitBarcodeDriverLicense.from(driverLicense: driverLicense)
        }

        return result
    }

    static func formatToString(_ format: BarcodeFormat) -> String {
        if format.contains(.code128) { return "CODE_128" }
        if format.contains(.code39) { return "CODE_39" }
        if format.contains(.code93) { return "CODE_93" }
        if format.contains(.codaBar) { return "CODABAR" }
        if format.contains(.dataMatrix) { return "DATA_MATRIX" }
        if format.contains(.EAN13) { return "EAN_13" }
        if format.contains(.EAN8) { return "EAN_8" }
        if format.contains(.ITF) { return "ITF" }
        if format.contains(.qrCode) { return "QR_CODE" }
        if format.contains(.UPCA) { return "UPC_A" }
        if format.contains(.UPCE) { return "UPC_E" }
        if format.contains(.PDF417) { return "PDF417" }
        if format.contains(.aztec) { return "AZTEC" }
        return "UNKNOWN"
    }

    static func valueTypeToString(_ valueType: BarcodeValueType) -> String {
        switch valueType {
        case .contactInfo: return "CONTACT_INFO"
        case .email: return "EMAIL"
        case .ISBN: return "ISBN"
        case .phone: return "PHONE"
        case .product: return "PRODUCT"
        case .SMS: return "SMS"
        case .text: return "TEXT"
        case .URL: return "URL"
        case .wiFi: return "WIFI"
        case .geographicCoordinates: return "GEO"
        case .calendarEvent: return "CALENDAR_EVENT"
        case .driversLicense: return "DRIVER_LICENSE"
        default: return "UNKNOWN"
        }
    }
}

// MARK: - Email

struct RNMLKitBarcodeEmail: Record {
    @Field
    var address: String?
    @Field
    var body: String?
    @Field
    var subject: String?
    @Field
    var type: String = "UNKNOWN"

    static func from(email: BarcodeEmail) -> RNMLKitBarcodeEmail {
        let result = RNMLKitBarcodeEmail()
        result.address = email.address
        result.body = email.body
        result.subject = email.subject
        result.type = emailTypeToString(email.type)
        return result
    }

    static func emailTypeToString(_ type: BarcodeEmailType) -> String {
        switch type {
        case .work: return "WORK"
        case .home: return "HOME"
        default: return "UNKNOWN"
        }
    }
}

// MARK: - Phone

struct RNMLKitBarcodePhone: Record {
    @Field
    var number: String?
    @Field
    var type: String = "UNKNOWN"

    static func from(phone: BarcodePhone) -> RNMLKitBarcodePhone {
        let result = RNMLKitBarcodePhone()
        result.number = phone.number
        result.type = phoneTypeToString(phone.type)
        return result
    }

    static func phoneTypeToString(_ type: BarcodePhoneType) -> String {
        switch type {
        case .work: return "WORK"
        case .home: return "HOME"
        case .fax: return "FAX"
        case .mobile: return "MOBILE"
        default: return "UNKNOWN"
        }
    }
}

// MARK: - SMS

struct RNMLKitBarcodeSMS: Record {
    @Field
    var message: String?
    @Field
    var phoneNumber: String?

    static func from(sms: BarcodeSMS) -> RNMLKitBarcodeSMS {
        let result = RNMLKitBarcodeSMS()
        result.message = sms.message
        result.phoneNumber = sms.phoneNumber
        return result
    }
}

// MARK: - URL Bookmark

struct RNMLKitBarcodeURLBookmark: Record {
    @Field
    var title: String?
    @Field
    var url: String?

    static func from(url: BarcodeURLBookmark) -> RNMLKitBarcodeURLBookmark {
        let result = RNMLKitBarcodeURLBookmark()
        result.title = url.title
        result.url = url.url
        return result
    }
}

// MARK: - WiFi

struct RNMLKitBarcodeWifi: Record {
    @Field
    var ssid: String?
    @Field
    var password: String?
    @Field
    var type: String = "OPEN"

    static func from(wifi: BarcodeWifi) -> RNMLKitBarcodeWifi {
        let result = RNMLKitBarcodeWifi()
        result.ssid = wifi.ssid
        result.password = wifi.password
        result.type = wifiEncryptionTypeToString(wifi.type)
        return result
    }

    static func wifiEncryptionTypeToString(_ type: BarcodeWiFiEncryptionType) -> String {
        switch type {
        case .open: return "OPEN"
        case .WPA: return "WPA"
        case .WEP: return "WEP"
        default: return "OPEN"
        }
    }
}

// MARK: - Geo Point

struct RNMLKitBarcodeGeoPoint: Record {
    @Field
    var latitude: Double = 0.0
    @Field
    var longitude: Double = 0.0

    static func from(geoPoint: BarcodeGeoPoint) -> RNMLKitBarcodeGeoPoint {
        let result = RNMLKitBarcodeGeoPoint()
        result.latitude = geoPoint.latitude
        result.longitude = geoPoint.longitude
        return result
    }
}

// MARK: - Address

struct RNMLKitBarcodeAddress: Record {
    @Field
    var addressLines: [String] = []
    @Field
    var type: String = "UNKNOWN"

    static func from(address: BarcodeAddress) -> RNMLKitBarcodeAddress {
        let result = RNMLKitBarcodeAddress()
        result.addressLines = address.addressLines ?? []
        result.type = addressTypeToString(address.type)
        return result
    }

    static func addressTypeToString(_ type: BarcodeAddressType) -> String {
        switch type {
        case .work: return "WORK"
        case .home: return "HOME"
        default: return "UNKNOWN"
        }
    }
}

// MARK: - Person Name

struct RNMLKitBarcodePersonName: Record {
    @Field
    var formattedName: String?
    @Field
    var first: String?
    @Field
    var last: String?
    @Field
    var middle: String?
    @Field
    var prefix: String?
    @Field
    var pronunciation: String?
    @Field
    var suffix: String?

    static func from(name: BarcodePersonName) -> RNMLKitBarcodePersonName {
        let result = RNMLKitBarcodePersonName()
        result.formattedName = name.formattedName
        result.first = name.first
        result.last = name.last
        result.middle = name.middle
        result.prefix = name.prefix
        result.pronunciation = name.pronunciation
        result.suffix = name.suffix
        return result
    }
}

// MARK: - Contact Info

struct RNMLKitBarcodeContactInfo: Record {
    @Field
    var addresses: [RNMLKitBarcodeAddress]?
    @Field
    var emails: [RNMLKitBarcodeEmail]?
    @Field
    var name: RNMLKitBarcodePersonName?
    @Field
    var phones: [RNMLKitBarcodePhone]?
    @Field
    var urls: [String]?
    @Field
    var jobTitle: String?
    @Field
    var organization: String?

    static func from(contactInfo: BarcodeContactInfo) -> RNMLKitBarcodeContactInfo {
        var result = RNMLKitBarcodeContactInfo()

        if let addresses = contactInfo.addresses {
            result.addresses = addresses.map { RNMLKitBarcodeAddress.from(address: $0) }
        }

        if let emails = contactInfo.emails {
            result.emails = emails.map { RNMLKitBarcodeEmail.from(email: $0) }
        }

        if let name = contactInfo.name {
            result.name = RNMLKitBarcodePersonName.from(name: name)
        }

        if let phones = contactInfo.phones {
            result.phones = phones.map { RNMLKitBarcodePhone.from(phone: $0) }
        }

        result.urls = contactInfo.urls
        result.jobTitle = contactInfo.jobTitle
        result.organization = contactInfo.organization

        return result
    }
}

// MARK: - Calendar Event

struct RNMLKitBarcodeCalendarEvent: Record {
    @Field
    var eventDescription: String?
    @Field
    var location: String?
    @Field
    var organizer: String?
    @Field
    var status: String?
    @Field
    var summary: String?
    @Field
    var start: String?
    @Field
    var end: String?

    static func from(calendarEvent: BarcodeCalendarEvent) -> RNMLKitBarcodeCalendarEvent {
        let result = RNMLKitBarcodeCalendarEvent()
        result.eventDescription = calendarEvent.eventDescription
        result.location = calendarEvent.location
        result.organizer = calendarEvent.organizer
        result.status = calendarEvent.status
        result.summary = calendarEvent.summary

        // Convert dates to ISO8601 strings if present
        if let start = calendarEvent.start {
            result.start = ISO8601DateFormatter().string(from: start)
        }
        if let end = calendarEvent.end {
            result.end = ISO8601DateFormatter().string(from: end)
        }

        return result
    }
}

// MARK: - Driver License

struct RNMLKitBarcodeDriverLicense: Record {
    @Field
    var addressCity: String?
    @Field
    var addressState: String?
    @Field
    var addressStreet: String?
    @Field
    var addressZip: String?
    @Field
    var birthDate: String?
    @Field
    var documentType: String?
    @Field
    var expiryDate: String?
    @Field
    var firstName: String?
    @Field
    var gender: String?
    @Field
    var issuingCountry: String?
    @Field
    var lastName: String?
    @Field
    var licenseNumber: String?
    @Field
    var middleName: String?

    static func from(driverLicense: BarcodeDriverLicense) -> RNMLKitBarcodeDriverLicense {
        var result = RNMLKitBarcodeDriverLicense()
        result.addressCity = driverLicense.addressCity
        result.addressState = driverLicense.addressState
        result.addressStreet = driverLicense.addressStreet
        result.addressZip = driverLicense.addressZip
        result.birthDate = driverLicense.birthDate
        result.documentType = driverLicense.documentType
        result.expiryDate = driverLicense.expiryDate
        result.firstName = driverLicense.firstName
        result.gender = driverLicense.gender
        result.issuingCountry = driverLicense.issuingCountry
        result.lastName = driverLicense.lastName
        result.licenseNumber = driverLicense.licenseNumber
        result.middleName = driverLicense.middleName

        return result
    }
}

// MARK: - Helper Types from Core

struct RNMLKitPoint: Record {
    @Field
    var x: CGFloat = 0
    @Field
    var y: CGFloat = 0

    static func fromNSValue(point: NSValue) -> RNMLKitPoint {
        let cgPoint = point.cgPointValue
        let result = RNMLKitPoint()
        result.x = cgPoint.x
        result.y = cgPoint.y
        return result
    }
}

struct RNMLKitRect: Record {
    @Field
    var origin: RNMLKitPoint = RNMLKitPoint()
    @Field
    var size: RNMLKitPoint = RNMLKitPoint()

    static func fromCGRect(rect: CGRect) -> RNMLKitRect {
        var result = RNMLKitRect()

        let origin = RNMLKitPoint()
        origin.x = rect.origin.x
        origin.y = rect.origin.y

        let size = RNMLKitPoint()
        size.x = rect.width
        size.y = rect.height

        result.origin = origin
        result.size = size

        return result
    }
}

File: modules/react-native-mlkit-barcode-scanning/android/src/main/java/red/infinite/reactnativemlkit/barcodescanning/RNMLKitBarcodeScannerResult.kt
package red.infinite.reactnativemlkit.barcodescanning

import android.graphics.Point
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.records.Field
import com.google.mlkit.vision.barcode.common.Barcode
import red.infinite.reactnativemlkit.core.RNMLKitRect
import red.infinite.reactnativemlkit.core.RNMLKitPoint

data class RNMLKitBarcodeScannerResultRecord(
    @Field var barcodes: List<RNMLKitBarcode> = mutableListOf(),
    @Field var imagePath: String
) : Record

class RNMLKitBarcodeScannerResult(private val barcodes: List<Barcode>, private val imagePath: String) {
    val record: RNMLKitBarcodeScannerResultRecord
        get() = RNMLKitBarcodeScannerResultRecord(
            barcodes = barcodes.map { barcode -> RNMLKitBarcode.fromBarcode(barcode) },
            imagePath = imagePath
        )
}

data class RNMLKitBarcode(
    @Field var frame: RNMLKitRect = RNMLKitRect.zero(),
    @Field var rawValue: String? = null,
    @Field var rawData: List<Int>? = null,
    @Field var displayValue: String? = null,
    @Field var format: String = "unknown",
    @Field var cornerPoints: List<RNMLKitPoint> = emptyList(),
    @Field var valueType: String = "UNKNOWN",
    @Field var email: RNMLKitBarcodeEmail? = null,
    @Field var phone: RNMLKitBarcodePhone? = null,
    @Field var sms: RNMLKitBarcodeSMS? = null,
    @Field var url: RNMLKitBarcodeURL? = null,
    @Field var wifi: RNMLKitBarcodeWifi? = null,
    @Field var geoPoint: RNMLKitBarcodeGeoPoint? = null,
    @Field var contactInfo: RNMLKitBarcodeContactInfo? = null,
    @Field var calendarEvent: RNMLKitBarcodeCalendarEvent? = null,
    @Field var driverLicense: RNMLKitBarcodeDriverLicense? = null
) : Record {
    companion object {
        fun fromBarcode(barcode: Barcode): RNMLKitBarcode {
            return RNMLKitBarcode(
                frame = barcode.boundingBox?.let { RNMLKitRect.fromRect(it) } ?: RNMLKitRect.zero(),
                rawValue = barcode.rawValue,
                rawData = barcode.rawBytes?.map { it.toInt() },
                displayValue = barcode.displayValue,
                format = formatIntToString(barcode.format),
                cornerPoints = barcode.cornerPoints?.map { point ->
                    RNMLKitPoint(point.x.toFloat(), point.y.toFloat())
                } ?: emptyList(),
                valueType = valueTypeIntToString(barcode.valueType),
                email = barcode.email?.let { RNMLKitBarcodeEmail.fromEmail(it) },
                phone = barcode.phone?.let { RNMLKitBarcodePhone.fromPhone(it) },
                sms = barcode.sms?.let { RNMLKitBarcodeSMS.fromSms(it) },
                url = barcode.url?.let { RNMLKitBarcodeURL.fromUrl(it) },
                wifi = barcode.wifi?.let { RNMLKitBarcodeWifi.fromWifi(it) },
                geoPoint = barcode.geoPoint?.let { RNMLKitBarcodeGeoPoint.fromGeoPoint(it) },
                contactInfo = barcode.contactInfo?.let { RNMLKitBarcodeContactInfo.fromContactInfo(it) },
                calendarEvent = barcode.calendarEvent?.let { RNMLKitBarcodeCalendarEvent.fromCalendarEvent(it) },
                driverLicense = barcode.driverLicense?.let { RNMLKitBarcodeDriverLicense.fromDriverLicense(it) }
            )
        }

        fun formatIntToString(format: Int): String {
            return when (format) {
                Barcode.FORMAT_ALL_FORMATS -> "all"
                Barcode.FORMAT_CODE_128 -> "code128"
                Barcode.FORMAT_CODE_39 -> "code39"
                Barcode.FORMAT_CODE_93 -> "code93"
                Barcode.FORMAT_CODABAR -> "codaBar"
                Barcode.FORMAT_DATA_MATRIX -> "dataMatrix"
                Barcode.FORMAT_EAN_13 -> "EAN13"
                Barcode.FORMAT_EAN_8 -> "EAN8"
                Barcode.FORMAT_ITF -> "ITF"
                Barcode.FORMAT_PDF417 -> "PDF417"
                Barcode.FORMAT_QR_CODE -> "qrCode"
                Barcode.FORMAT_UPC_A -> "UPCA"
                Barcode.FORMAT_UPC_E -> "UPCE"
                Barcode.FORMAT_AZTEC -> "AZTEC"
                else -> "unknown"
            }
        }

        fun valueTypeIntToString(valueType: Int): String {
            return when (valueType) {
                Barcode.TYPE_UNKNOWN -> "UNKNOWN"
                Barcode.TYPE_CONTACT_INFO -> "CONTACT_INFO"
                Barcode.TYPE_EMAIL -> "EMAIL"
                Barcode.TYPE_ISBN -> "ISBN"
                Barcode.TYPE_PHONE -> "PHONE"
                Barcode.TYPE_PRODUCT -> "PRODUCT"
                Barcode.TYPE_SMS -> "SMS"
                Barcode.TYPE_TEXT -> "TEXT"
                Barcode.TYPE_URL -> "URL"
                Barcode.TYPE_WIFI -> "WIFI"
                Barcode.TYPE_GEO -> "GEO"
                Barcode.TYPE_DRIVER_LICENSE -> "DRIVER_LICENSE"
                Barcode.TYPE_CALENDAR_EVENT -> "CALENDAR_EVENT"
                else -> "UNKNOWN"
            }
        }
    }
}

data class RNMLKitBarcodeEmail(
    @Field var address: String? = null,
    @Field var body: String? = null,
    @Field var subject: String? = null,
    @Field var type: String = "UNKNOWN"
) : Record {
    companion object {
        fun fromEmail(email: Barcode.Email): RNMLKitBarcodeEmail {
            return RNMLKitBarcodeEmail(
                address = email.address,
                body = email.body,
                subject = email.subject,
                type = when (email.type) {
                    Barcode.Email.TYPE_HOME -> "HOME"
                    Barcode.Email.TYPE_WORK -> "WORK"
                    else -> "UNKNOWN"
                }
            )
        }
    }
}

data class RNMLKitBarcodePhone(
    @Field var number: String? = null,
    @Field var type: String = "UNKNOWN"
) : Record {
    companion object {
        fun fromPhone(phone: Barcode.Phone): RNMLKitBarcodePhone {
            return RNMLKitBarcodePhone(
                number = phone.number,
                type = when (phone.type) {
                    Barcode.Phone.TYPE_HOME -> "HOME"
                    Barcode.Phone.TYPE_WORK -> "WORK"
                    Barcode.Phone.TYPE_FAX -> "FAX"
                    Barcode.Phone.TYPE_MOBILE -> "MOBILE"
                    else -> "UNKNOWN"
                }
            )
        }
    }
}

data class RNMLKitBarcodeSMS(
    @Field var message: String? = null,
    @Field var phoneNumber: String? = null
) : Record {
    companion object {
        fun fromSms(sms: Barcode.Sms): RNMLKitBarcodeSMS {
            return RNMLKitBarcodeSMS(
                message = sms.message,
                phoneNumber = sms.phoneNumber
            )
        }
    }
}

data class RNMLKitBarcodeURL(
    @Field var title: String? = null,
    @Field var url: String? = null
) : Record {
    companion object {
        fun fromUrl(url: Barcode.UrlBookmark): RNMLKitBarcodeURL {
            return RNMLKitBarcodeURL(
                title = url.title,
                url = url.url
            )
        }
    }
}

data class RNMLKitBarcodeWifi(
    @Field var ssid: String? = null,
    @Field var password: String? = null,
    @Field var type: String = "OPEN"
) : Record {
    companion object {
        fun fromWifi(wifi: Barcode.WiFi): RNMLKitBarcodeWifi {
            return RNMLKitBarcodeWifi(
                ssid = wifi.ssid,
                password = wifi.password,
                type = when (wifi.encryptionType) {
                    Barcode.WiFi.TYPE_WPA -> "WPA"
                    Barcode.WiFi.TYPE_WEP -> "WEP"
                    else -> "OPEN"
                }
            )
        }
    }
}

data class RNMLKitBarcodeGeoPoint(
    @Field var latitude: Double = 0.0,
    @Field var longitude: Double = 0.0
) : Record {
    companion object {
        fun fromGeoPoint(geoPoint: Barcode.GeoPoint): RNMLKitBarcodeGeoPoint {
            return RNMLKitBarcodeGeoPoint(
                latitude = geoPoint.lat,
                longitude = geoPoint.lng
            )
        }
    }
}

data class RNMLKitBarcodeAddress(
    @Field var addressLines: List<String>? = null,
    @Field var type: String = "UNKNOWN"
) : Record {
    companion object {
        fun fromAddress(address: Barcode.Address): RNMLKitBarcodeAddress {
            return RNMLKitBarcodeAddress(
                addressLines = address.addressLines?.toList(),
                type = when (address.type) {
                    Barcode.Address.TYPE_HOME -> "HOME"
                    Barcode.Address.TYPE_WORK -> "WORK"
                    else -> "UNKNOWN"
                }
            )
        }
    }
}

data class RNMLKitBarcodePersonName(
    @Field var formattedName: String? = null,
    @Field var first: String? = null,
    @Field var last: String? = null,
    @Field var middle: String? = null,
    @Field var prefix: String? = null,
    @Field var pronunciation: String? = null,
    @Field var suffix: String? = null
) : Record {
    companion object {
        fun fromPersonName(name: Barcode.PersonName): RNMLKitBarcodePersonName {
            return RNMLKitBarcodePersonName(
                formattedName = name.formattedName,
                first = name.first,
                last = name.last,
                middle = name.middle,
                prefix = name.prefix,
                pronunciation = name.pronunciation,
                suffix = name.suffix
            )
        }
    }
}

data class RNMLKitBarcodeContactInfo(
    @Field var addresses: List<RNMLKitBarcodeAddress>? = null,
    @Field var emails: List<RNMLKitBarcodeEmail>? = null,
    @Field var name: RNMLKitBarcodePersonName? = null,
    @Field var phones: List<RNMLKitBarcodePhone>? = null,
    @Field var urls: List<String>? = null,
    @Field var jobTitle: String? = null,
    @Field var organization: String? = null
) : Record {
    companion object {
        fun fromContactInfo(contactInfo: Barcode.ContactInfo): RNMLKitBarcodeContactInfo {
            return RNMLKitBarcodeContactInfo(
                addresses = contactInfo.addresses?.map { RNMLKitBarcodeAddress.fromAddress(it) },
                emails = contactInfo.emails?.map { RNMLKitBarcodeEmail.fromEmail(it) },
                name = contactInfo.name?.let { RNMLKitBarcodePersonName.fromPersonName(it) },
                phones = contactInfo.phones?.map { RNMLKitBarcodePhone.fromPhone(it) },
                urls = contactInfo.urls?.toList(),
                jobTitle = contactInfo.title,
                organization = contactInfo.organization
            )
        }
    }
}

data class RNMLKitBarcodeCalendarEvent(
    @Field var eventDescription: String? = null,
    @Field var location: String? = null,
    @Field var organizer: String? = null,
    @Field var status: String? = null,
    @Field var summary: String? = null,
    @Field var start: Long? = null,
    @Field var end: Long? = null
) : Record {
    companion object {
        fun fromCalendarEvent(event: Barcode.CalendarEvent): RNMLKitBarcodeCalendarEvent {
            return RNMLKitBarcodeCalendarEvent(
                eventDescription = event.description,
                location = event.location,
                organizer = event.organizer,
                status = event.status,
                summary = event.summary,
                start = event.start?.let { calendarDateTimeToMillis(it) },
                end = event.end?.let { calendarDateTimeToMillis(it) }
            )
        }

        private fun calendarDateTimeToMillis(dateTime: Barcode.CalendarDateTime): Long {
            val calendar = java.util.Calendar.getInstance()
            calendar.set(dateTime.year, dateTime.month - 1, dateTime.day, dateTime.hours, dateTime.minutes, dateTime.seconds)
            return calendar.timeInMillis
        }
    }
}

data class RNMLKitBarcodeDriverLicense(
    @Field var firstName: String? = null,
    @Field var middleName: String? = null,
    @Field var lastName: String? = null,
    @Field var gender: String? = null,
    @Field var addressCity: String? = null,
    @Field var addressState: String? = null,
    @Field var addressStreet: String? = null,
    @Field var addressZip: String? = null,
    @Field var birthDate: String? = null,
    @Field var documentType: String? = null,
    @Field var expiryDate: String? = null,
    @Field var issuingDate: String? = null,
    @Field var issuingCountry: String? = null
) : Record {
    companion object {
        fun fromDriverLicense(license: Barcode.DriverLicense): RNMLKitBarcodeDriverLicense {
            return RNMLKitBarcodeDriverLicense(
                firstName = license.firstName,
                middleName = license.middleName,
                lastName = license.lastName,
                gender = license.gender,
                addressCity = license.addressCity,
                addressState = license.addressState,
                addressStreet = license.addressStreet,
                addressZip = license.addressZip,
                birthDate = license.birthDate,
                documentType = license.documentType,
                expiryDate = license.expiryDate,
                issuingDate = license.issueDate,
                issuingCountry = license.issuingCountry
            )
        }
    }
}
